### PR TITLE
WIP: http: Use libhtp-rs v8

### PR DESCRIPTION
**Pull request by cccs-rtmorti**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/2696) ticket: 2696

Describe changes:
- Use libhtp-rs.
- Rebased from #9031

Rebased from v7. Did not tackle the commit separation yet - will wait until the functionality is considered good enough to merge.

LIBHTP_REPO=https://github.com/cccs-rtmorti/libhtp-rs
LIBHTP_BRANCH=main

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -112,3 +112,6 @@ pub mod plugin;
 pub mod lzma;
 pub mod util;
 pub mod ffi;
+
+//Re-export htp symbols
+pub use htp::c_api::*;
```
Comment (catenacyber) [inline rust/src/lib.rs:117]: Can you re explain why we need this ?

File: src/app-layer-htp-file.c [hunk 2]
```diff
@@ -83,9 +83,9 @@ int HTPFileOpen(HtpState *s, HtpTxUserData *tx, const uint8_t *filename, uint16_
  * @param[in] rawvalue
  * @param[out] range
  *
- * @return HTP_OK on success, HTP_ERROR on failure.
+ * @return HTP_STATUS_OK on success, HTP_STATUS_ERROR on failure.
  */
-int HTPParseContentRange(bstr *rawvalue, HTTPContentRange *range)
+int HTPParseContentRange(const bstr *rawvalue, HTTPContentRange *range)
```
Comment (catenacyber) [inline src/app-layer-htp-file.c:88]: adding `const` may be a separate commit

File: src/app-layer-htp.h [hunk 3]
```diff
@@ -45,8 +45,7 @@
 #define HTP_CONFIG_DEFAULT_REQUEST_INSPECT_WINDOW       4096U
 #define HTP_CONFIG_DEFAULT_RESPONSE_INSPECT_MIN_SIZE    32768U
 #define HTP_CONFIG_DEFAULT_RESPONSE_INSPECT_WINDOW      4096U
-#define HTP_CONFIG_DEFAULT_FIELD_LIMIT_SOFT             9000U
```
Comment (catenacyber) [inline src/app-layer-htp.h:48]: Was it unused ?

File: src/app-layer-htp.h [hunk 4]
```diff
@@ -76,66 +75,15 @@ enum {
...
-
     /* suricata errors/warnings */
-    HTTP_DECODER_EVENT_MULTIPART_GENERIC_ERROR,
-    HTTP_DECODER_EVENT_MULTIPART_NO_FILEDATA,
-    HTTP_DECODER_EVENT_MULTIPART_INVALID_HEADER,
-
-    HTTP_DECODER_EVENT_TOO_MANY_WARNINGS,
-
-    HTTP_DECODER_EVENT_FAILED_PROTOCOL_CHANGE,
+    HTTP_DECODER_EVENT_MULTIPART_GENERIC_ERROR = 200,
```
Comment (catenacyber) [inline src/app-layer-htp.h:79]: Why were most removed ?

**Comment by catenacyber** [comment]
I guess it would be nice to have a script doing all the automatic changes to see what's left after that...

File: rust/src/lib.rs (hunk 1 above)
Comment (cccs-rtmorti) [reply rust/src/lib.rs:117]: See previous discussion: [https://github.com/OISF/suricata/pull/8335#discussion_r1063245087](here).

Short story, the htp symbols need to be exported in the rust .so so suricata can find them.

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (cccs-rtmorti) [reply src/app-layer-htp-file.c:88]: Understood. There was a [previous discussion](https://github.com/OISF/suricata/pull/8335#discussion_r1063252451) on this as well. 

I would like to save the commit refactoring for after everything else is done. It's actually quite a lot of work to try to separate the commits and keep everything compiling at the same time, and doing it every time for v8, v9, v10, ... means redoing it every time.

File: src/app-layer-htp.h (hunk 3 above)
Comment (cccs-rtmorti) [reply src/app-layer-htp.h:48]: Yes. libhtp never implemented soft limits so the soft limits were never actually used. This is mentioned in the commit message as well:

> htp_config_set_field_limits(soft_limit, hard_limit) changed to
> htp_config_set_field_limit(limit) because libhtp didn't implement soft
> limits.

**Comment by cccs-rtmorti** [comment]
> I guess it would be nice to have a script doing all the automatic changes to see what's left after that...

I can separate the function / constant / const changes out into separate commits, but they won't compile. If it would make your life easier I am happy to do that just for review purposes.

File: src/app-layer-htp.h (hunk 4 above)
Comment (cccs-rtmorti) [reply src/app-layer-htp.h:79]: Also there was a [previous discussion](https://github.com/OISF/suricata/pull/8335#discussion_r1063261726).

File: rust/src/lib.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/lib.rs:117]: Thanks

**Comment by catenacyber** [comment]
> I can separate the function / constant / const changes out into separate commits, but they won't compile.

Oh, why is that ?

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (cccs-rtmorti) [reply src/app-layer-htp-file.c:88]: I am not sure what you mean by prerequisite commits? You mean just the const changes? Everything else actually requires the library to be present.

**Comment by cccs-rtmorti** [comment]
> > I can separate the function / constant / const changes out into separate commits, but they won't compile.
> 
> Oh, why is that ?

Because if I have one commit where I just change the function names but not the constants then it will not compile, because the constant names will still reference libhtp-c, and libhtp-c won't be there any more?

Suppose there are 4 separate commits. 
- One removes libhtp and replaces it with libhtp-rs from configure, etc
- One changes all the constant names
- One changes all the function names / signatures (includes adding const, because the new APIs return const)
- One modifies suricata to conform to any API changes (eg: log message handling) and make actual functional changes

None of these commits will compile individually. The first commit will not compile because libhtp-c will be gone, but all the code in suricata still references the libhtp-c API. The second commit will not compile because the suricata will still be using libhtp-c function names and structures. The third commit will not compile because there are some APIs in libhtp-c that are not present in libhtp-rs. The last commit will compile because all of the required changes will be there to switch from libhtp-c to libhtp-rs.

I am under the impression that if a PR is segmented into multiple commits, each of those commits must - by themselves - also compile, and that this is important for bisection? This is what [this QA check](https://github.com/OISF/suricata/blob/master/.github/workflows/commits.yml) is for, yes?

So anyway, I can segment all the commits out so all of the mechanical changes are in separate commits, but this will just be for review purposes, since that QA check will fail. If you want every commit to build on it's own, then all of the changes to switch to the new library and conform to the different API will have to be in a single commit, no?

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-htp-file.c:88]: I indeed mean everything that does not need the new library, (but that the new library may require)

Also removal of the unused soft limit value ?
You know better what else there may be :-)

**Comment by catenacyber** [comment]
Ok, but I meant a script that does all automatic changes in one go.

This is not important :
As a reviewer (and maybe as a PR author who needs to rebase it), it is easier to revise the script and check that it does the diff in the commit, than check all the lines which got changed

**Comment by cccs-rtmorti** [comment]
> Ok, but I meant a script that does all automatic changes in one go.
> 
> This is not important : As a reviewer (and maybe as a PR author who needs to rebase it), it is easier to revise the script and check that it does the diff in the commit, than check all the lines which got changed

Yes, I understand you want a script that does this, but that script does not exist. If you just don't want to look at the mechanical changes I can separate them out.

If you want a script I can try to write a bunch of sed, but I doubt it will be perfect.

**Comment by catenacyber** [comment]
> If you want a script I can try to write a bunch of sed, but I doubt it will be perfect.

Ok for me, do not worry about this script :)

**Comment by catenacyber** [comment]
So, overall I am good with this.

What is missing before making it a not-draft PR ?
(rebase for newer changes, some differential fuzzing stuff, wait for branch 8 to open...)

@victorjulien should I take this over ?

**Comment by cccs-rtmorti** [comment]
> What is missing before making it a not-draft PR ? (rebase for newer changes, some differential fuzzing stuff, wait for branch 8 to open...)

I think that's it, and there is some private tests that apparently failed that were mentioned on a previous PR version. I don't have access to those tests, but am happy to try to triage given the nature of the failures.

**Comment by catenacyber** [comment]
Rebased in https://github.com/OISF/suricata/pull/9609
